# Question
**\*** *How many of the popular Spotify 2023 songs are above vs under 120 BPM? What was the lowest and Highest BPM? What was the median?*

## Study
*I will figure this out by conducting a observational study. I will be doing a observational study due to only having data from the Spotify popular songs 2023 list.*

In [49]:
import matplotlib.pyplot as plt
import seaborn as sns
import pandas as pd

In [50]:
df = pd.read_csv(r'C:\Users\*******\!jupyter\data\spotify-2023.csv', encoding='latin-1') # latin-1 fixes encoding bug with the data because some songs have weird characters in the name
bpm = df['bpm']

In [51]:
sns.set_theme('paper')
sns.set_style('darkgrid')

o120 = 0 # over
u120 = 0 # under
i120 = 0 # is

for i in range(len(bpm)):
  if int(bpm[i]) > 120:
    o120 += 1
  elif int(bpm[i]) < 120:
    u120 += 1
  elif int(bpm[i]) == 120:
    i120 += 1
  else:
    continue

def showgraph(): # not calling it cause its in the conclusion but I wanna keep the code in here
  sns.histplot(df, color="#ff9a2a", x="bpm")
  plt.axvline(bpm.median(), color="#ff4545", linestyle="--")
  plt.text(
    bpm.median(), 0.95,
    f"Median: {bpm.median():.0f}",
    transform=plt.gca().get_xaxis_transform(),
    ha="left",
    va="top"
  )
  plt.axvline(bpm.min(), color="#8645ff", linestyle="--")
  plt.text(
    bpm.min(), 0.95,
    f"Min BPM: {bpm.min():.0f}",
    transform=plt.gca().get_xaxis_transform(),
    ha="left",
    va="top"
  )
  plt.axvline(bpm.max(), color="#4599ff", linestyle="--")
  plt.text(
    bpm.max(), 0.95,
    f"Max BPM: {bpm.max():.0f}",
    transform=plt.gca().get_xaxis_transform(),
    ha="left",
    va="top"
  )
  plt.show()

print(f"above: {o120} | under: {u120} | is: {i120} | total: {o120 + u120 + i120}")
print(f"median: {bpm.median():.2f} | mean: {bpm.mean():.2f} | quantile: {bpm.quantile():.2f} | min: {bpm.min()} | max: {bpm.max()}")

above: 477 | under: 437 | is: 39 | total: 953
median: 121.00 | mean: 122.54 | quantile: 121.00 | min: 65 | max: 206


# Conclusion
In conclusion, out of a total of 953 songs, there are more songs above 120 BPM *(477)* than songs under 120 BPM *(437)* or songs that are 120 BPM *(39)*. The highest BPM in the data was 206 while the lowest BPM was 65. The median of the data was 121.

**This graph helps visualize the difference in the BPM of the songs:**<br>
<img src="https://raw.githubusercontent.com/hxproxy/AI-Class/refs/heads/main/Semester1/Unit1/spotify/spotify_graph_2_from_python.png" width="536" height="341" alt="graph"><br>
The graph shows that the most common bpm is roughly ~121 BPM with a median of 121.